File: java/binary_to_decimal_converter/binary_to_decimal_converter_9.java
/**
 * This program is designed to convert a binary number (base-2) to a decimal number (base-10).
 * The purpose of this program is to help programmers understand the process of converting
 * binary numbers to decimal numbers through a step-by-step approach. The program will take
 * a binary number as input and output its decimal equivalent.
 */

import java.util.Scanner;

public class BinaryToDecimalConverter {

    // Function to convert binary to decimal
    public static int convertBinaryToDecimal(String binary) {
        int decimal = 0;
        int length = binary.length();
        int power = 1; // Initialize power of 2

        // Loop through each character in the binary string
        for (int i = length - 1; i >= 0; i--) {
            char bit = binary.charAt(i);
            if (bit == '1') {
                decimal += power;
            }
            power *= 2; // Update power of 2
        }

        return decimal;
    }

    // Main function to execute the program
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        // Prompt the user to enter a binary number
        System.out.println("Enter a binary number:");
        String binaryInput = scanner.nextLine();

        // Convert the binary number to decimal
        int decimalOutput = convertBinaryToDecimal(binaryInput);

        // Display the decimal equivalent
        System.out.println("The decimal equivalent of binary number " + binaryInput + " is: " + decimalOutput);

        // Close the scanner
        scanner.close();
    }
}

File: perl/binary_to_decimal_converter/binary_to_decimal_converter_9.pl
# This Perl program is designed to convert a binary number to its decimal equivalent.
# The program takes a binary number as input from the user and processes it to produce
# the corresponding decimal number. This is a useful exercise for understanding how
# binary numbers are converted to decimal numbers and for practicing Perl programming.

use strict;
use warnings;

# Function to convert binary to decimal
sub binary_to_decimal {
    my ($binary) = @_;
    my $decimal = 0;
    my $length = length($binary);
    my $power = 0;

    # Loop through each digit in the binary number
    for (my $i = $length - 1; $i >= 0; $i--) {
        my $digit = substr($binary, $i, 1);
        if ($digit eq '1') {
            $decimal += 2 ** $power;
        }
        $power++;
    }

    return $decimal;
}

# Main program
print "Enter a binary number: ";
my $binary_input = <STDIN>;
chomp($binary_input);

# Check if the input is a valid binary number
if ($binary_input =~ /^[01]+$/) {
    my $result = binary_to_decimal($binary_input);
    print "The decimal equivalent of $binary_input is $result\n";
} else {
    print "Invalid input. Please enter a valid binary number.\n";
}

# Additional unused variables and functions
my $sunny = "sunny";
my $rainy = "rainy";
my $cloudy = "cloudy";

sub unused_function {
    my $unused_variable = "This is not used";
    return $unused_variable;
}

# End of the program
